Sources/PackageRegistry/RegistryDownloadsManager.swift

```diff
@@ -0,0 +1,310 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import Foundation
+import PackageModel
+import TSCBasic
+import PackageLoading
+
+public class RegistryDownloadsManager {
+    public typealias Delegate = RegistryDownloadsManagerDelegate
+
+    private let fileSystem: FileSystem
+    private let path: AbsolutePath
+    private let cachePath: AbsolutePath?
+    private let registryClient: RegistryClient
+    private let checksumAlgorithm: HashAlgorithm
+    private let delegate: Delegate?
+
+    private var pendingLookups = [PackageIdentity: DispatchGroup]()
+    private var pendingLookupsLock = NSLock()
+
+    public init(
+        fileSystem: FileSystem,
+        path: AbsolutePath,
+        cachePath: AbsolutePath?,
+        registryClient: RegistryClient,
+        checksumAlgorithm: HashAlgorithm,
+        delegate: Delegate?
+    ) {
+        self.fileSystem = fileSystem
+        self.path = path
+        self.cachePath = cachePath
+        self.registryClient = registryClient
+        self.checksumAlgorithm = checksumAlgorithm
+        self.delegate = delegate
+    }
+
+    public func lookup(
+        package: PackageIdentity,
+        version: Version,
+        observabilityScope: ObservabilityScope,
+        delegateQueue: DispatchQueue,
+        callbackQueue: DispatchQueue,
+        completion: @escaping  (Result<AbsolutePath, Error>) -> Void
+    ) {
+        // wrap the callback in the requested queue
+        let completion = { result in callbackQueue.async { completion(result) } }
+        
+        let packageRelativePath: RelativePath!
```

Comment:
It's unfortunate that Swift can't reason about this being set in all code paths that don't return.